### DISCUSSION TITLE
Open-Source Business Central Apps Contribution Framework

### DISCUSSION THREAD

**Discussion by vody** [microsoft/alguidelines#243] [General]
Dear Business Central Community,

I hope you're all doing great! I'm working on a new open-source Business Central app. As I dive into this project, I’ve been anticipating several challenges that I'd like to discuss with the community:
- How do I share my app with a broader community?
- How do I build community trust in my app?
- Where can I learn best practices for developing open-source BC apps?
- What open-source license should I use?
- How do I find others who can help with design or offer advice?
- If our app is still just an idea, where can I validate that its use case will benefit the community?

These challenges are common for many open-source app creators, and it's crucial for us as a community to provide a framework to help resolve them.

I was thinking—what if we had a central place for all our open-source BC apps? Imagine a GitHub repo within the Microsoft organisation (like BCTech) where we can all collaborate. This could be a mono-repo with all community-owned open-source apps as [submodules](https://git-scm.com/book/en/v2/Git-Tools-Submodules), serving as a hub for:
- Sharing and discussing new app ideas (using idea-type issues).
- Learning and spreading best practices (using discussions).
- Providing a single NuGet package feed for all open-source apps (using packages).
- Open-source app creators will form a working group to support newbies and each other on the open-source journey.

Each app would still reference the creator as a publisher, keeping the creator's branding. AppSource publishing will also be done by the creator. However, hosting the code within a Microsoft organisation could help build trust and ensure we follow good coding practices.

I’d love to hear what you all think about this idea. How do you see us supporting open-source BC app creators and making this community even stronger? Your thoughts and ideas would be super valuable!
Looking forward to hearing from you all!

Tagging all BC open-source app creators I found on GitHub. Please mention anyone else who may bring value to this discussion. 
@picazin, @NovoaDev, @ajkauffmann, @Bertverbeek4PS, @DmitryKatson, @waldo1001, @hougaard, @ethan-sorenson, @TheDoubleH, @JeremyVyska, @joneug, @juliandittmann, @kine, @mynavblog, @msnraju, @OlegLobakov, @StefanMaron, @demiliani, @vjekob, @Drakonian

**Comment by StefanMaron** [comment]
I like everything except having all apps in one repo.
I can not explain exactly why, but it doesn't feel right.

Maybe just a place on MSLearn to reference all the different apps, similar to what I created in the past?

https://stefanmaron.onrender.com/opensourcealprojects/

**Reply by vody** [reply]
@StefanMaron, it will be a references. It's a submodules: https://git-scm.com/book/en/v2/Git-Tools-Submodules

**Comment by Drakonian** [comment]
Every open-source contributor asks these questions, and it's completely normal. There are no guarantees when developing such projects from the outset. We just create applications and hope they will be useful to the community. We also hope that people will learn about them.

Having one repository through git submodules within the Microsoft organization can work well, but only if the git submodules automatically pull changes from the source repos. I haven't used them in a long time and don't remember if this is the case. If it is, then it is a pretty good idea. In this case, it is much more convenient if the development takes place in the source repository.

This could help more people learn about open-source projects and encourage others to create their own open-source projects.

However, there are questions that arise with this approach:
- Who will be responsible for selecting projects for this list?
- What are the selection criteria for such a list?
- Will there be additional requirements for open-source code?
- Who will be responsible for moderating discussions in this repository? Is discussion even necessary in such a repository?

Honestly I think there is one more thing Microsoft can do to make life easier for open-source projects. It is more transparent and simple conditions for placing such **open-source free** projects in Microsoft App Source.

I personally find that I can't put my free apps on App Source because of the complicated procedure and the requirement to have a company.

**Reply by vody** [reply]
@Drakonian Thank you for your feedback. 

> Having one repository through git submodules within the Microsoft organization can work well, but only if the git submodules automatically pull changes from the source repos. I haven't used them in a long time and don't remember if this is the case. If it is, then it is a pretty good idea.

Git submodules make commit references, so in case of a new commit in the submodule release(main) branch, we would need to have a small automation which the submodule can kick off to update a commit reference. 

> In this case, it is much more convenient if the development takes place in the source repository.

Agreed. We discussed it with @JeremyVyska, and this approach will require minimal effort from the creator. We can keep a full app commit history. Of course, because each app will have its own repository, any app-specific issues will be seated in an app repository and not in a "hub" one.

> Who will be responsible for selecting projects for this list? What are the selection criteria for such a list?

I imagine we as a community will draft a set of requirements that include the basics (like having a readme, dedicated range, publisher, repo issues, owner, etc.). The app creator initiates a PR. Existing app creators can validate compliance with the requirements and then approve a PR. 

> Will there be additional requirements for open-source code?

@Drakonian, can you please elaborate on what additional requirements are required for open-source code? Are you for an AL code quality? 

> Who will be responsible for moderating discussions in this repository? Is discussion even necessary in such a repository?

We are quite a healthy community :) so I'm not sure we need extensive moderation. @JeremyVyska, can you answer on how much moderation you do for [alguidelines discussions](https://github.com/microsoft/alguidelines/discussions) and our [Discord Server](https://discord.gg/sw8jsJNg). I think we would need a discussions to make sure we communicate on BC open-source issues and challenges and are able to build more help materials for creators.

> Honestly, I think there is one more thing Microsoft can do to make life easier for open-source projects. It is more transparent and simple conditions for placing such open-source, free projects in Microsoft AppSource.

This repo can be a perfect place to discuss some of these conditions. Microsoft will likely listen if it does not violate their commercial marketplace policies. 

> I personally find that I can't put my free apps on App Source because of the complicated procedure and the requirement to have a company.

I know that some people tagged in this discussion went through the process of creating such an entity, and they will be happy to share knowledge and help others.
If creating your own entity is off the table, having such a forum will definitely benefit creators like you, as it will help you find an entity that will be happy to work with you on your app and publish it for you.

**Reply by Drakonian** [reply]
> @Drakonian, can you please elaborate on what additional requirements are required for open-source code? Are you for an AL code quality?
> 
> >
Including but not limited to code requirements. The goal is to avoid creating an artificial barrier that would be difficult to overcome, so projects outside the list do not become "third-rate."

Unfortunately, barriers are everywhere, and better do not to create another one. For example:
- The barrier to publishing an application in App Source.
- The barrier to System/Base contribution (high project compilation complexity, poor tools).
- The barrier of being a Microsoft partner. If you are not a partner, the issues you find on the platform are of no interest to anyone. If you are not a partner, you cannot work fully with clients.
- The BC Yammer barrier. You cannot access Yammer with a personal account, and with a work account, you will wait more than six months for a response! Personally, with my current work account, I get an error when entering the email. (Two years without BC Yammer, lol).
- BC Ideas barrier. At least for now we can implement some ideas as open-source contribution, but a lot of ideas related to platform and process is very slow.

These are just the barriers I remembered, and there are many more.


> Git submodules make commit references, so in case of a new commit in the submodule release(main) branch, we would need to have a small automation which the submodule can kick off to update a commit reference.

Yes, if there is some automation, then it should work.

> I know that some people tagged in this discussion went through the process of creating such an entity, and they will be happy to share knowledge and help others.
> If creating your own entity is off the table, having such a forum will definitely benefit creators like you, as it will help you find an entity that will be happy to work with you on your app and publish it for you.

As Stefan would say, "I can not explain exactly why, but it doesn't feel right" when a third party publishes your free open-source project in App Source :)


Overall, I support this initiative and hope it will help creators rather than restrict them.

**Reply by vody** [reply]
@Drakonian thank you for your support. I will be keen to work with you on this framework.

> Including but not limited to code requirements. The goal is to avoid creating an artificial barrier that would be difficult to overcome, so projects outside the list do not become "third-rate." Unfortunately, barriers are everywhere, and better do not to create another one. For example: ...

Thank you for pointing this out. I fully agree and personally think that none of the listed ones should become a barrier to joining a list. 

> As Stefan would say, "I can not explain exactly why, but it doesn't feel right" when a third party publishes your free open-source project in App Source :)

@StefanMaron can correct me, but I think his comment was around code hosting in a single repo. I pointed out that "AppSource publishing will also be done by the creator," so there will be no other publisher except the creator. Only in cases where the creator wants but cannot publish an app with his entity, someone within a community of creators/publishers can help with it. But I think this will not be a one-off exercise, but rather a joined future collaboration between the original creator and publisher.

**Reply by Drakonian** [reply]
> @StefanMaron can correct me, but I think his comment was around code hosting in a single repo. I pointed out that "AppSource publishing will also be done by the creator," so there will be no other publisher except the creator. Only in cases where the creator wants but cannot publish an app with his entity, someone within a community of creators/publishers can help with it. But I think this will not be a one-off exercise, but rather a joined future collaboration between the original creator and publisher.

Of course, the phrase was about something else, but it fits perfectly as a response in this case also 😄  

I understand your idea, but I'm still not sure that  involving third-party publishers is a good solution to this problem.


> @Drakonian thank you for your support. I will be keen to work with you on this framework.

Thanks for initiative, I will be happy to help in any way I can. Open-source is very important, and any development in this direction will always have my support.

**Reply by JeremyVyska** [reply]
> > In this case, it is much more convenient if the development takes place in the source repository.
> 
> Agreed. We discussed it with @JeremyVyska, and this approach will require minimal effort from the creator. We can keep a full app commit history. Of course, because each app will have its own repository, any app-specific issues will be seated in an app repository and not in a "hub" one.

That was partly the thought process.  A Submodule link can behave as a Psuedo "Catalog" of projects.  Then each repo linked can maintain their own Issues, templates, alerts, etc.   If a creator wants all that, cool, make a repo, and PR a submodule to it.  If someone just wants to make a little thing and contribute to a central place, this gives those a home.   It felt the best approach to Invite/Involve without creating too many of those extra barriers.

> > Who will be responsible for moderating discussions in this repository? Is discussion even necessary in such a repository?
> 
> We are quite a healthy community :) so I'm not sure we need extensive moderation. @JeremyVyska, can you answer on how much moderation you do for [alguidelines discussions](https://github.com/microsoft/alguidelines/discussions) and our [Discord Server](https://discord.gg/sw8jsJNg). I think we would need a discussions to make sure we communicate on BC open-source issues and challenges and are able to build more help materials for creators.

It's not a lot, but I will admit it takes good community people to be active to make it feel like a thriving community.

**Comment by DmitryKatson** [comment]
@vody I had the same idea several months ago. But after some thinking, i got into conclusion than developing OOS Free AppSource app independently, make no sense. If i want this app to be widely used by the community, be free and open-sourced - the best way is to contribute to the base/system app directly. 

- First, you get immediate understanding of how valuable your app is going to be
- Second, you keep all the history of your commits - if you develop it as open source from the beginning
- Third, you get "free" code reviews from Microsoft directly
- Last, you a free from your app support. Microsoft will handle that.

For me, this appeared the better way to go. 
Example: https://github.com/microsoft/BCApps/pulls?q=is%3Apr+is%3Aclosed+author%3ADmitryKatson

**Reply by Drakonian** [reply]
I can't agree with this. This approach imposes many restrictions. Not every free open-source application is suitable for the Base/System app.

In fact, I have several such examples myself:
- [Data Editor for BC](https://github.com/Drakonian/data-editor-for-bc)
- [BC PDF Viewer](https://github.com/Drakonian/bc-pdf-viewer)
- [Drag and Drop Factbox](https://github.com/Drakonian/drag-and-drop-factbox)

I'm sure there are many such "unsuitable" projects from other open-source creators as well. With this approach, you are simply denying these projects the right to exist.

Your approach is not a replacement for open-source projects, but merely an addition.

**Reply by vody** [reply]
@DmitryKatson, I agree with @Drakonian that there are classes of apps that do not go to the Base/System app. 
But this is not a discussion of what should exist and where, but to help BC open-source apps gain more popularity and support. I'm pretty sure some app ideas can be initially created as open-source separate apps, then get to our repo, and then transition to a Base/System app if Microsoft is happy to push it there.

**Comment by hougaard** [comment]
There are also a few practical challenges, like object number series, that have to be different between PTE deployment and AppSource, like dependencies to Microsoft localizations (or can open source only be W1 dependent) etc....

I fail to see any value in an organized open-source-bc-apps-repo outside of system app, I agree with @DmitryKatson

**Reply by vody** [reply]
@hougaard, with an initial framework defined, I am sure we, as a community, can provide guidelines on resolving such challenges. Some ideas: 
- Object number series for AppSource: Each publisher (creator) has a range from Microsoft (same as now). 
- Object number series for PTE (may not be many): overlaps are acceptable (as it's unlikely overlapping apps need to be in the same box), but we may come to a point where we will suggest using "AL Object ID Ninja" as a tool to ensure PTE apps do not overlap so they can be installed in the same environment. 
- Dependencies to Microsoft localizations: I don't see why we should suggest any restrictions on this point, but I'm happy to discuss. 

In terms of value. I agree with @Drakonian point in [this comment](https://github.com/microsoft/alguidelines/discussions/243#discussioncomment-10219245). Moreover, some apps are not supposed to be in Base/System: 
- [OpenFeature-al](https://github.com/thetanz/OpenFeature-al) is an app primarily created for other ISVs and using ApplicationArea outside of "correct" purpose. 
- [AdminHub](https://github.com/thetanz/AdminHub-al) (not yet open-sourced) is a tool for VAR's to manage tenants and apps. It has nothing to do with the finance system.

**Reply by hougaard** [reply]
So, who is the consumer of this? How does this end up in a tenant as published apps?

**Reply by Drakonian** [reply]
> So, who is the consumer of this? How does this end up in a tenant as published apps?

Consumers are the target audience for such projects. Nothing unusual about that.

PTE extensions are used by many, and this is normal. There are many users who install and work with free open-source PTE without any issues. In my experience, it is not just consultants and developers, but also end-users who do this.

**Reply by hougaard** [reply]
... but what's the difference form the current situation, why would the end-user be better off with this?

**Reply by Drakonian** [reply]
> ... but what's the difference form the current situation, why would the end-user be better off with this?

That's a very good question, and I'll try to answer it from my perspective. Perhaps @vody will have a different response.

Such a repository would be more well-known than each individual project, which in turn would increase the number of users who utilize these projects. Additionally, it would stimulate others to create similar projects, seeing the growing interest from both users and Microsoft.

Those who want to create such projects will always be able to get help from an experienced community in this place, which will also increase the number of these projects. Furthermore, the quality and security of these projects will improve due to more rigorous control from the community.

As a result, we will have more open-source projects, more interested users, and higher quality.

The user gets a greater selection of quality open-source projects that solve their problems, and this is the benefit for the user.

**Comment by juliandittmann** [comment]
Hi @vody ,

thank you for bringing us together here on this important topic. 
I am glad that you bring it up once more since i faced a lot of your challanges 3 years ago.

I will answer your questions first.

**Q: How do I share my app with a broader community?**
A: You already shared it on your GitHub profile. This is a great start. In my opinion, open source apps from private individuals are not widespread and therefore not necessarily known to developers/partners or end customers.


**Q:  How do I build community trust in my app?**
A: Building trust in an app is a long process and it will take time. Especially in the ERP context.
Be reliable, be honest, communicate clearly, listen, offer help,  take small steps, make decisions carefully.

**Q: Where can I learn best practices for developing open-source BC apps?**
A: When asking about developer best practices i would recommend the resources microsoft provides - Microsoft also provides us with the compiler :). When asking about open source i think there is no recipe, only great examples. I learn by trying, by failing and by discussions.

**Q: What open-source license should I use?**
A:  It's completely up to you what you want to achieve. Open source means that it is open, but there may still be conditions attached to it. MIT License is my way to go.

**Q: How do I find others who can help with design or offer advice?**
A: Here or the Discord I guess :).

**Q: If our app is still just an idea, where can I validate that its use case will benefit the community?**
A: There is no validation. The customer/user decides. Once I created an app with some collegues to play DOOM in Business Central for fun.

**Now the key Question: Do we need a central place for all our open-source BC apps?** 
A: Not in general, but I think so as a first start up point. Three years ago i had quite some time to think about open source for BC.

### **Why is open source very important?**

For me it is raising the bar.

For a long time, the business central/navision world was extremely sealed off due to a developer license. Also documentation and important resources were shared on blogs  and forums or even behind paywalls. This has changed we have much more access now!! Thank you to everyone getting us to this point <3 

So I started also sharing.

In the BC world we are still strongly dependent on Microsoft. Therefore i think an open source movement without Microsoft's approval will fail in the longterm. 
So I showed Microsoft at BCTechDays my simple projects and asked what they think about it. The response was very positive and that I should keep going. 

My conclusion is that Microsoft is not against open source and is even willing to help. Also Microsoft has their own ideas on open source. The topic this is not new.

In the end I created a github company https://github.com/OpenBusinessCentralApps. 
I setup up CICD with AL-GO for Github, an Azure backend, Application Insights for the apps and even a website https://openbcapps.com/ to which you can pullrequest.

But I failed to maintain those projects.


**What i can offer to help the community:**

I can administrate and share access to OpenBusinessCentralApps with the community.

You can contact me: juliandittmann@openbcapps.com
